### ISSUE TITLE
Retina Display 3x scale factor on iPhone 7 and 8 Plus creating rounding errors in layout

### ISSUE THREAD

**Issue by jasorod** [facebook/yoga#661]
I've noticed an odd layout issue on iPhone 8 Plus and iPhone 7 Plus devices that have a 
 3x screen scale size, and a pointScaleFactor set to 0.33333, leaving a fractional value of 0.3333 when calling the floating point modulo of the pointScaleFactor inside of `YGRoundValueToPixelGrid` on any given whole-number value.

Let's say we have a simple view hierarchy 3 levels deep. Our code in a UIViewController would look like the following:

```
override func loadView() {
    self.view = UIView()
    self.view.frame = UIScreen.main.bounds
    self.view.translatesAutoresizingMaskIntoConstraints = false
}

override func viewDidLoad() {
    super.viewDidLoad()

    self.view.configureLayout{ (layout) in
      layout.isEnabled = true
      layout.width = YGValue(self.view.bounds.size.width)
      layout.height = YGValue(self.view.bounds.size.height)
      layout.alignItems = .center
      layout.justifyContent = .center
    }

    self.view.backgroundColor = .white

    let contentView = UIView()
    contentView.backgroundColor = .lightGray
    contentView.configureLayout { (layout) in
      layout.isEnabled = true
      layout.width = 320
      layout.height = 80
    }
    self.view.addSubview(contentView)

    let child1 = UIView()
    child1.backgroundColor = .red
    child1.alpha = 0.5
    child1.configureLayout{ (layout)  in
      layout.isEnabled = true
      layout.width = 100%
      layout.height = 100%
    }
    contentView.addSubview(child1)

    self.view.yoga.applyLayout(preservingOrigin: true)
}
```

One would expect, since `child1` is set to 100% of the width and height of its 320x80 point parent container UIView that its dimensions would also be 320x80 points, and this is the case on devices like an iPhone 7 or 8 where the floating point modulo of the `pointScaleFactor` and the dimensional value being rounded in `YGRoundValueToPixelGrid` is zero.  But on an iPhone 7 or 8 Plus where the `pointScaleFactor` is set to 0.333333, the result is a dimension of 320.333333333333x80.3333333333333 while the parent is still 320x80.  So the child is actually **larger** than the parent that it's suppose to contain after `YGRoundValueToPixelGrid` is called.

This rounding error accumulates with each new call to `applyLayout` (i.e., when the device is rotated).  For instance, after 10 rotations the frame size of `child1` becomes 322x82, a couple points larger than the parent container UIView frame.

Since I'm new to the codebase, I'm not fully aware of all the implications associated with changing `YGRoundValueToPixelGrid`, but wouldn't a proper assumption be that no matter what the `pointScaleFactor` of a device is, any whole number value passed to `YGRoundValueToPixelGrid`  should be a no-op?  In other words there is no need to round a whole-number value to the pixel grid since it should already align to the grid, it's only fractional values need to be rounded to the pixel grid.

I can create a pull request with a patch if that assumption is correct.

**Comment by jasorod** [comment]
Looks like this issue was fixed in commit b378a685a434926cc22a5477a7e764023b066478. Thanks.

**Comment by woehrl01** [comment]
If you have any more rounding problems, please feel free to reopen.
